### PR TITLE
rbd: add exclusive-lock and journaling image features for rbd image

### PR DESCRIPTION
**Pull request by Madhu-1**
The current rbd only supports the layering feature for the rbd image. Add exclusive-lock and journaling image features for the rbd.

Signed-off-by: woohhan <woohyung_han@tmax.co.kr>
Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>

This will abandon the original PR
Closes: #684
updates  #676

### CODE REVIEW

**Comment by Madhu-1** [comment]
added not-ready-to-review so that I need to test it out first.

**Comment by Madhu-1** [comment]
@nixpanic any idea how to load `nbd` modules to minikube VM?

File: internal/rbd/rbd_util.go [hunk 1]
```diff
@@ -707,25 +727,17 @@ func genVolFromVolumeOptions(ctx context.Context, volOptions, credentials map[st
...
 	}
 
-	util.ExtendedLog(ctx, "setting disableInUseChecks on rbd volume to: %v", disableInUseChecks)
 	rbdVol.DisableInUseChecks = disableInUseChecks
 
-	rbdVol.Mounter, ok = volOptions["mounter"]
-	if !ok {
-		rbdVol.Mounter = rbdDefaultMounter
-	}
+	util.ExtendedLog(ctx, "setting disableInUseChecks: %t imagefeatures: %v mounter: %s", disableInUseChecks, rbdVol.imageFeatureSet, rbdVol.Mounter)
```
Comment (nixpanic) [inline internal/rbd/rbd_util.go]: I think you need to log `rbdVol.imageFeatureSet.Names()` to get some useful results.

File: internal/rbd/rbd_util.go [hunk 2]
```diff
@@ -135,8 +136,27 @@ type rbdSnapshot struct {
 	RequestName    string
 }
 
+// imageFeature represents required image features and value.
+type imageFeature struct {
+	// needRbdNbd indicates whether this image feature requires an rbd-nbd mounter
+	needRbdNbd bool
```
Comment (nixpanic) [inline internal/rbd/rbd_util.go:149]: Are these features really only available when using rbd-nbd? Or does it depend on the kernel version? @dillaman

File: internal/rbd/rbd_util.go (hunk 1 above)
Comment (Madhu-1) [reply internal/rbd/rbd_util.go]: done

**Comment by humblec** [comment]
@Madhu-1 as discussed deferring this to 3.2.0 :+1:

**Comment by Madhu-1** [comment]
@humblec @nixpanic PTAL

**Comment by Madhu-1** [comment]
@mergifyio rebase

File: internal/rbd/rbd_util.go (hunk 2 above)
Comment (dillaman) [reply internal/rbd/rbd_util.go:149]: Journaling is never supported with the kernel. You would have to go pretty far back for a kernel that doesn't support layering / exclusive-lock. The object-map/fast-diff/deep-flatten features are another story. It's all documented here [1] but your mileage may vary when dealing w/ RHEL/CentOS due to backporting.

[1] https://docs.ceph.com/en/latest/rbd/rbd-config-ref/#image-features

File: internal/rbd/rbd_util_test.go [hunk 3]
```diff
@@ -42,3 +44,86 @@ func TestHasSnapshotFeature(t *testing.T) {
...
+		{
+			"layering,exclusive-loc,journaling",
+			&rbdVolume{
+				Mounter: rbdNbdMounter,
+			},
+			true,
+			"invalid feature exclusive-loc",
+		},
+		{
+			"ayering",
```
Comment (dillaman) [inline internal/rbd/rbd_util_test.go:111]:
```suggestion
			"layering",
```

Comment (nixpanic) [reply internal/rbd/rbd_util_test.go:111]: This is intentional, an error is expected

**Comment by nixpanic** [comment]
@Madhu-1, this needs a rebase...

**Comment by salanki** [comment]
Is it time to merge this now that kernel support is prevalent?

File: docs/deploy-rbd.md [hunk 4]
```diff
@@ -48,19 +48,19 @@ make image-cephcsi
...
+| `volumeNamePrefix`                                                                                  | no                   | Prefix to use for naming RBD images (defaults to `csi-vol-`).                                                                                                                                                                                             |
+| `snapshotNamePrefix`                                                                                | no                   | Prefix to use for naming RBD snapshot images (defaults to `csi-snap-`).                                                                                                                                                                                   |
+| `imageFeatures`                                                                                     | no                   | RBD image features. CSI RBD currently supports `layering`, `journaling`, `exclusive-lock` features. If `journaling` is enabled, must enable `exclusive-lock` too. See [man pages](http://docs.ceph.com/docs/mimic/man/8/rbd/#cmdoption-rbd-image-feature) |
```
Comment (dillaman) [inline docs/deploy-rbd.md]: https://docs.ceph.com/en/latest/man/8/rbd/...

**Comment by Madhu-1** [comment]
@humblec @nixpanic PTAL

File: internal/rbd/rbd_util.go [hunk 5]
```diff
@@ -916,6 +926,29 @@ func genVolFromVolumeOptions(ctx context.Context, volOptions, credentials map[st
...
+	featureSet := sets.NewString(arr...)
+	for _, f := range arr {
+		sf, found := supportedFeatures[f]
+		if !found {
+			return fmt.Errorf("invalid feature %s", f)
+		}
+
+		for _, r := range sf.dependsOn {
+			if !featureSet.Has(r) {
+				return fmt.Errorf("feature %s requires %s", f, r)
```
Comment (humblec) [inline internal/rbd/rbd_util.go]: may be a better wording : `feature %s requires %s to be set` ?

File: internal/rbd/rbd_util.go [hunk 6]
```diff
@@ -885,29 +905,19 @@ func genVolFromVolumeOptions(ctx context.Context, volOptions, credentials map[st
...
-			}
-		}
-		rbdVol.imageFeatureSet = librbd.FeatureSetFromNames(arr)
+	if err = rbdVol.validateImageFeatures(volOptions["imageFeatures"]); err != nil {
+		util.ErrorLog(ctx, "failed to validate image features %v", err)
+		return nil, err
 	}
 
-	util.ExtendedLog(ctx, "setting disableInUseChecks on rbd volume to: %v", disableInUseChecks)
+	util.ExtendedLog(ctx, "setting disableInUseChecks: %t imagefeatures: %v mounter: %s", disableInUseChecks, rbdVol.imageFeatureSet.Names(), rbdVol.Mounter)
```
Comment (humblec) [inline internal/rbd/rbd_util.go]: nit: space between image and features

**Comment by humblec** [comment]
@Madhu-1 looks like other PR merge cause a conflict.

**Comment by Madhu-1** [comment]
> @Madhu-1 looks like other PR merge cause a conflict.

@humblec looks like we merged the Newly created PR first. That would cause a lot of issues(resolve merge conflict and rebase) to the PR's which are pending review for a long time.

**Comment by humblec** [comment]
> > @Madhu-1 looks like other PR merge cause a conflict.
> 
> @humblec looks like we merged the Newly created PR first. That would cause a lot of issues(resolve merge conflict and rebase) to the PR's which are pending review for a long time.

ahh, :(.. we could have added `DNM` with that..

**Comment by Madhu-1** [comment]
> > > @Madhu-1 looks like other PR merge cause a conflict.
> > 
> > 
> > @humblec looks like we merged the Newly created PR first. That would cause a lot of issues(resolve merge conflict and rebase) to the PR's which are pending review for a long time.
> 
> ahh, :(.. we could have added `DNM` with that..

I was expecting old PR's will get reviewed first

**Review by humblec — DISMISSED** [review]

**Comment by Madhu-1** [comment]
>Warning: Permanently added 'n62.dusty.ci.centos.org,172.19.2.126' (ECDSA) to the list of known hosts.
Error from server: etcdserver: request timed out

E2E failure

**Comment by humblec** [comment]
@nixpanic ptal ?

File: e2e/rbd.go [hunk 7]
```diff
@@ -467,6 +467,18 @@ var _ = Describe("RBD", func() {
 				}
 			})
 
+			// TODO: enable this test when we support rbd-nbd mounter in E2E.
+			// nbd module should be present on the host machine to run use the
+			// rbd-nbd mounter.
```
Comment (nixpanic) [inline e2e/rbd.go:472]: @pkalever FYI. Once this PR is merged, and you have rebased your rbd-nd one, you can enable this test as well.

File: examples/rbd/storageclass.yaml [hunk 8]
```diff
@@ -33,7 +33,8 @@ parameters:
    # creation (thin provisioning is the default).
    thickProvision: "false"
    # (required) RBD image features, CSI creates image with image-format 2
-   # CSI RBD currently supports only `layering` feature.
+   # CSI RBD currently supports `layering`, `journaling`, `exclusive-lock`
+   # features. If `journaling` is enabled, must enable `exclusive-lock` too.
```
Comment (nixpanic) [inline examples/rbd/storageclass.yaml:37]: This should be a comma separated list, without spaces, I think? Might be good to have an example, or mention it explicitly.

Comment (Madhu-1) [reply examples/rbd/storageclass.yaml:37]: Yes, it should be comma separate without space.

**Review by nixpanic — APPROVED** [review]

**Comment by nixpanic** [comment]
> /retest ci/centos/upgrade-tests-cephfs

etcdserver timeout while deploying rook

**Review by humblec — APPROVED** [review]

**Pull request merged**